featureflags.go

```diff
@@ -94,91 +141,454 @@ func (poller *FeatureFlagsPoller) fetchNewFeatureFlags() {
+
+		result, err = poller.getFeatureFlagVariant(featureFlag, flagConfig.Key, flagConfig.DistinctId)
+		if err != nil {
+			return nil, nil
+		}
 	}
-	return false, nil
+
+	return result, err
 }
 
-func (poller *FeatureFlagsPoller) GetFeatureFlag(key string, distinctId string, defaultResult interface{}) (interface{}, error) {
+func (poller *FeatureFlagsPoller) GetAllFlags(flagConfig FeatureFlagPayloadNoKey) (map[string]interface{}, error) {
+	response := map[string]interface{}{}
 	featureFlags := poller.GetFeatureFlags()
+	fallbackToDecide := false
 
-	if len(featureFlags) < 1 {
-		return defaultResult, nil
+	if len(featureFlags) == 0 {
+		fallbackToDecide = true
+	} else {
+		for _, storedFlag := range featureFlags {
+			result, err := poller.computeFlagLocally(storedFlag, flagConfig.DistinctId, flagConfig.Groups, flagConfig.PersonProperties, flagConfig.GroupProperties)
+			if err != nil {
+				poller.Errorf("Unable to compute flag locally - %s", err)
+				fallbackToDecide = true
+			} else {
+				response[storedFlag.Key] = result
+			}
+		}
 	}
 
-	featureFlag := FeatureFlag{Key: ""}
+	if fallbackToDecide && !flagConfig.OnlyEvaluateLocally {
+		result, err := poller.getFeatureFlagVariants(flagConfig.DistinctId, flagConfig.Groups)
 
-	// avoid using flag for conflicts with Golang's stdlib `flag`
-	for _, storedFlag := range featureFlags {
-		if key == storedFlag.Key {
-			featureFlag = storedFlag
-			break
+		if err != nil {
+			return response, err
+		} else {
+			for k, v := range result {
+				response[k] = v
+			}
+		}
+	}
+
+	return response, nil
+}
+
+func (poller *FeatureFlagsPoller) computeFlagLocally(flag FeatureFlag, distinctId string, groups Groups, personProperties Properties, groupProperties map[string]Properties) (interface{}, error) {
+	if flag.EnsureExperienceContinuity != nil && *flag.EnsureExperienceContinuity {
+		return nil, &InconclusiveMatchError{"Flag has experience continuity enabled"}
+	}
+
+	if !flag.Active {
+		return false, nil
+	}
+
+	if flag.Filters.AggregationGroupTypeIndex != nil {
+
+		groupName, exists := poller.groups[fmt.Sprintf("%d", *flag.Filters.AggregationGroupTypeIndex)]
+
+		if !exists {
+			errMessage := "Flag has unknown group type index"
+			return nil, errors.New(errMessage)
+		}
+
+		_, exists = groups[groupName]
+
+		if !exists {
+			errMessage := fmt.Sprintf("FEATURE FLAGS] Can't compute group feature flag: %s without group names passed in", flag.Key)
+			return nil, errors.New(errMessage)
+		}
+
+		focusedGroupProperties := groupProperties[groupName]
+		return matchFeatureFlagProperties(flag, distinctId, focusedGroupProperties)
+	} else {
+		return matchFeatureFlagProperties(flag, distinctId, personProperties)
+	}
+}
+
+func getMatchingVariant(flag FeatureFlag, distinctId string) (interface{}, error) {
+	lookupTable := getVariantLookupTable(flag)
+
+	for _, variant := range lookupTable {
+		minHash, err := _hash(flag.Key, distinctId, "variant")
```

Comment:
minor nit: min and maxHash can be calculated outside the loop, so we need to calc them only once